service/rpc/share.go

```diff
@@ -76,3 +62,104 @@ func (h *Handler) handleSharesByNamespaceRequest(w http.ResponseWriter, r *http.
 		log.Errorw("serving request", "endpoint", namespacedSharesEndpoint, "err", err)
 	}
 }
+
+func (h *Handler) handleDataByNamespaceRequest(w http.ResponseWriter, r *http.Request) {
+	height, nID, err := parseGetByNamespaceArgs(r)
+	if err != nil {
+		writeError(w, http.StatusBadRequest, namespacedDataEndpoint, err)
+		return
+	}
+	shares, headerHeight, err := h.getShares(r.Context(), height, nID)
+	if err != nil {
+		writeError(w, http.StatusInternalServerError, namespacedDataEndpoint, err)
+		return
+	}
+	data, err := dataFromShares(shares)
+	if err != nil {
+		writeError(w, http.StatusInternalServerError, namespacedDataEndpoint, err)
+		return
+	}
+	resp, err := json.Marshal(&NamespacedDataResponse{
+		Data:   data,
+		Height: uint64(headerHeight),
+	})
+	if err != nil {
+		writeError(w, http.StatusInternalServerError, namespacedDataEndpoint, err)
+		return
+	}
+	_, err = w.Write(resp)
+	if err != nil {
+		log.Errorw("serving request", "endpoint", namespacedDataEndpoint, "err", err)
+	}
+}
+
+func (h *Handler) getShares(ctx context.Context, height uint64, nID namespace.ID) ([]share.Share, int64, error) {
+	// get header
+	var (
+		err    error
+		header *header.ExtendedHeader
+	)
+	switch height {
+	case 0:
+		header, err = h.header.Head(ctx)
+	default:
+		header, err = h.header.GetByHeight(ctx, height)
+	}
+	if err != nil {
+		return nil, 0, err
+	}
+	// perform request
+	shares, err := h.share.GetSharesByNamespace(ctx, header.DAH, nID)
+	return shares, header.Height, err
+}
+
+func dataFromShares(shares []share.Share) ([][]byte, error) {
+	for i := range shares {
+		insertDelimiter(shares[i])
+	}
+	messages, err := types.ParseMsgs(shares)
+	if err != nil {
+		return nil, err
+	}
+	data := make([][]byte, len(messages.MessagesList))
+	for i := range messages.MessagesList {
+		data[i] = messages.MessagesList[i].Data
+	}
+	return data, nil
+}
+
+func insertDelimiter(share share.Share) share.Share {
+	lenBuf := make([]byte, binary.MaxVarintLen64)
+	length := uint64(len(share))
+	n := binary.PutUvarint(lenBuf, length)
+
+	var withDelim []byte
+	if len(share)+n <= cap(share) {
+		withDelim = share[:cap(share)]
+	} else {
+		withDelim = make([]byte, len(share)+n)
+		copy(withDelim[:ipld.NamespaceSize], share[:ipld.NamespaceSize])
+	}
```

Comment:
LMAO, it's `realloc`, but in Go :D

While debugging, I noticed, that usually `cap(share)` is significantly bigger than  `len(share)` (`share` is a result of multiple `append`s, so it's reallocated few times, and has some space on the end).